Makes lighting more responsive.

MrsTonedOne opened the pull request:
Lowers priority, doubles tick rate, makes it run sooner in the tick. (bonus points, if it gets paused by MC_TICK_CHECK before it finishes, and there ends up being spare time in the byond tick left over, it will get some more runtime at the tail end of the tick, meaning it runs twice a tick during low lag times.)

I was avoiding doing this when i first made stoned mc because i wasn't sure if we'd hit some point where these get ticker subsystems get contentious, but that hasn't happened and now it seems like making this a ticker subsystem is sane way of doing this.

:cl:
experiment: Lighting was made more responsive.
/:cl:

MrsTonedOne committed:
Makes lighting more responsive.

Tacolizard commented:
with my pr, #22754, lighting actually moves to turfs slightly BEFORE you do. How exactly could this increase responsiveness. or am i thinking of something else/am dumb

MrStonedOne commented:
You are missing something.

Lighting doesn't "move", it updates the tiles as something else moves.

The key difference is the actual light or lack of light is static, it never moves, its a "overlay" on the tile that stays with the tile. when something relating to lighting changes, it requests a re-calculation for that last caster, this happens for things casting light moving, or the amount of light they cast to be changed.

The actual re-calculation happens in the lighting subsystem, ran by the MasterController every n 1/10 of a second, with that number before being 2, so 2/10s of a second

Lighting, like all subsystems, will be given a quota of how long it can run each fire, if it can't update everything in that time frame, it stops and is resumed later by the mc, this controlled by the priority value.

So the process goes:

Mob moves  
moved() calls the recalculation on lighting if the mob is emitting light.  
it queues an update with the lighting subsystem (sslighting),  
at some point the MasterController calls sslighting's fire().  
sslighting runs thru the queue of light sources (the thing that actually emits light) calling light_source's check()  
this calls light_source's remove_effect, removing the luminosity that light added from the turf for every turf the light was effecting on the old location
the turf(s) detecting this queues an update with sslighting to redraw it's lighting  
check() then calls add_effect to add its light to the new location, this will go thru every turf in range of the new location and adds to that turf's luminosity based on light strength and how far away that turf is  
the turf(s) detecting this queues an update with sslighting to redraw it's lighting
sslighting then runs thru the queue of turfs to update and calls redraw_lighting on the turfs.

You changed how turfs redrew their lighting so they didn't animate the change, I changed how often sslighting fires and how the mc allocates how long it has each fire.

We queue the changes up because a lot of times, the same light source or the same tile will move a bunch in on game tick, and it made no sense to re-calculate lighting changes that would never get sent out to clients because they would get re-calculated again later.

Queuing them up also allows the MC to allocate processing time to the lighting updates based on how loaded the server is, and what else needs to run.
